src/java.base/share/classes/java/lang/Module.java

```diff
@@ -256,25 +258,73 @@ public ModuleLayer getLayer() {
     /**
      * Update this module to allow access to restricted methods.
      */
-    Module implAddEnableNativeAccess() {
+    synchronized Module implAddEnableNativeAccess() {
         enableNativeAccess = true;
         return this;
     }
 
     /**
-     * Update all unnamed modules to allow access to restricted methods.
+     * Returns {@code true} if this module can access
+     * <a href="foreign/package-summary.html#restricted"><em>restricted</em></a> methods.
+     *
+     * @return {@code true} if this module can access <em>restricted</em> methods.
      */
-    static void implAddEnableNativeAccessAllUnnamed() {
-        ALL_UNNAMED_MODULE.enableNativeAccess = true;
+    @PreviewFeature(feature=PreviewFeature.Feature.FOREIGN)
+    public boolean isNativeAccessEnabled() {
+        Module target = enableNativeAccessHolder();
+        synchronized(target) {
+            return target.enableNativeAccess;
+        }
+    }
+
+    // Returns the Module object holds the enableNativeAccess
+    // flag for this module.
+    private Module enableNativeAccessHolder() {
+        return isNamed()? this : ALL_UNNAMED_MODULE;
+    }
+
+    // This is invoked from Reflection.ensureNativeAccess
+    void ensureNativeAccess(Class<?> owner, String methodName) {
+        // The target module whose enableNativeAccess flag is ensured
+        Module target = enableNativeAccessHolder();
+        // racy read of the enable native access flag
+        boolean isNativeAccessEnabled = target.enableNativeAccess;
+        if (!isNativeAccessEnabled) {
+            synchronized(target) {
+                // safe read of the enableNativeAccess of the target module
+                isNativeAccessEnabled = target.enableNativeAccess;
+                // check again with the safely read flag
+                if (isNativeAccessEnabled) {
+                   return;
```

Comment:
I'd suggest adding a comment like "another thread beat us to it - nothing to do"